### PR TITLE
Fix: Fallback manual copy for update_localdb()

### PR DESCRIPTION
**Pull request by iguessthislldo**
I originally started writing this as an issue I was going to open, but decided to look into it and found that I could fix it.

When using `tldr --update`, it exits with error code 1 and prints something like `Error: Could Not Rename: /tmp/tldrqquWXv/tldr-master to /home/fred/.tldrc/tldr-master/`. I compiled from latest commit on Arch Linux. It was shows the message when querying a command for the first time, but it also shows the tldr page, and no error code. After .tldrc has been created it only shows it for `tldr --update`.

This happens because `rename()` has an EXDEV (cross device) error. It can't copy the updated pages from /tmp to home because Arch mounts /tmp using tmpfs by default. I changed it so that only when `rename()` has a EXDEV error, it falls back to a recursive copy instead.

### CODE REVIEW

File: src/utils.c [hunk 1]
```diff
@@ -202,3 +202,163 @@ unzip(char const *path, char const *outpath)
...
+    if (n == -1) {
+        fprintf(stderr,
+            "Read error during copy from \"%s\".\n",
+            src
+        );
+        return 1;
+    }
+
+    /* Close source file */
+    if (close(src_fd) == -1) {
```
Comment (Leandros) [inline src/utils.c]: If this fails, the destination file is kept open.

File: src/utils.c [hunk 2]
```diff
@@ -202,3 +202,163 @@ unzip(char const *path, char const *outpath)
...
+        fprintf(stderr,
+            "Error: Couldn't open source file for copying: %s\n",
+            src
+        );
+        return 1;
+    }
+
+    /* Open destination file with the source's permissions.
+     */
+    if ((dest_fd = open(dest, dest_flags, perms)) == -1) {
```
Comment (Leandros) [inline src/utils.c]: If this fails the source file is kept open.

File: src/utils.c [hunk 3]
```diff
@@ -202,3 +202,163 @@ unzip(char const *path, char const *outpath)
...
+            return 1;
+        }
+    }
+
+    if (n == -1) {
+        fprintf(stderr,
+            "Read error during copy from \"%s\".\n",
+            src
+        );
+        return 1;
```
Comment (Leandros) [inline src/utils.c]: If this fails, destination and source files are kept open.

File: src/utils.c [hunk 4]
```diff
@@ -202,3 +202,163 @@ unzip(char const *path, char const *outpath)
...
+        fprintf(stderr,
+            "Error: Couldn't open destination file for copying: %s\n",
+            dest
+        );
+        return 1;
+    }
+    
+    /* Write to dest while we can read from src */
+    while ((n = read(src_fd, buffer, BUFSIZ)) > 0) {
+        if (write(dest_fd, buffer, (size_t) n) == -1) {
```
Comment (Leandros) [inline src/utils.c]: If this fails, destination and source files are kept open.

File: src/utils.c [hunk 5]
```diff
@@ -202,3 +202,163 @@ unzip(char const *path, char const *outpath)
...
+                }
+                /* Append relative path to destbuf, overwriting anything else
+                 * that might be there after the destination path.
+                 */
+                strncpy(destbuf + dest_len, cur->fts_path + src_len,
+                        cur->fts_pathlen - src_len); 
+                destbuf[dest_len + cur->fts_pathlen - src_len] = '\0';
+
+                if (cur_is_dir) { /* Create Directory with perms of src one */
+                    if (mkdir(destbuf, cur->fts_statp->st_mode)) {
```
Comment (Leandros) [inline src/utils.c]: If this fails,`fts_close()` is not called

File: src/utils.c [hunk 6]
```diff
@@ -202,3 +202,163 @@ unzip(char const *path, char const *outpath)
...
+                        if (errno != EEXIST) { /* Ignore if already exists */
+                            fprintf(stderr, "Couldn't create directory: %s\n",
+                                    destbuf);
+                            return 1;
+                        }
+                    }
+                    cur_is_dir = 0;
+                } else { /* Copy File */
+                    if (copyfile(cur->fts_path, destbuf,
```
Comment (Leandros) [inline src/utils.c]: If this fails,`fts_close()` is not called

File: src/local.c [hunk 7]
```diff
@@ -201,9 +201,21 @@ update_localdb(int verbose)
...
+        if (errno == EXDEV) {
+            /* tmp and outhome are on different file systems, do full copy
+             * and rm instead.
+             */
+            if(copytree(tmp, outhome)) {
+                fprintf(stderr, "Error: Could Not Move: %s to %s\n",
+                        tmp, outhome);
+                return 1;
+            }
+            /*rm(tmp);*/
```
Comment (iguessthislldo) [inline src/local.c]: Yeah, I thought I had to remove the directory for the move but the directory gets removed in the next if block with its parent outpath. I don't know why I commented it out, but either way it's useless and I'm going to remove it.

**Comment by iguessthislldo** [comment]
I think I addressed all your comments, also I hope you didn't mind that I copy and pasted the fts loop from the rm function. It was tempting and I'm not super familiar with unix apis, but I hope to be.

**Comment by LukasRypl** [comment]
Would be nice if this PR is accepted. Is there anything more what needs to be done?

Maybe a better solution would be to create a temporary directory inside `~/.tldrc/` instead of `/tmp` so that there is no need to handle renaming between different file systems.

**Comment by iguessthislldo** [comment]
Yeah, that seems like a more elegant solution than my attempt to recreate cp -r. Maybe I should have brought this up as an issue first after all, but I was a little excited to make my first contribution to a open source project.

I went ahead and changed it to how you described.

**Comment by iguessthislldo** [comment]
I just realized that @LukasRypl isn't @Leandros... I'm sorry I keep jumping the gun on what I'm doing here. @Leandros when you get back to this, I'm happy to go with whatever, as long as the problem is fixed, because I like program.

**Comment by Leandros** [comment]
Don't worry. I'll take a look as soon as I can.

**Comment by 12345ieee** [comment]
I hit this error, as well.

While I wait for the fix to be released, what is the workaround?

**Comment by 12345ieee** [comment]
For anybody landing here, this fixes the problem:

```
mkdir ~/.tldrc
cd ~/.tldrc
wget https://github.com/tldr-pages/tldr/archive/master.zip
unzip master.zip
rm master.zip
```

Script based on a cursory look at the code.

**Comment by gingerbeardman** [comment]
Above gets me working but does not fix the core issue. Eventually the app thinks the data is too old and I need to do the manual fix again.

**Comment by devhell** [comment]
@Leandros can this be merged please?

**Comment by bl-ue** [comment]
I haven't seen this PR.  I wonder if we can merge it. 🤔

**Comment by bl-ue** [comment]
Hey @iguessthislldo! I realize that it's been a long time, but would you be interested in working with me to merge this PR? I'm a tldr-pages maintainer so I can do that ;)

**Comment by iguessthislldo** [comment]
@bl-ue I moved on to a different tldr client a long time ago, but it would be nice if my first ever PR was merged. Looks like I deleted my fork at some point though, so I can't just simply reopen it.... I will see if I can get it set back up if I have some time.

**Comment by lwpie** [comment]
Any updates?

**Pull request closed without merging**